secure/src/globals/mod.rs

```diff
@@ -105,3 +112,14 @@ pub trait Hasher {
     ///
     fn hash_reset(&mut self, slice: &[u8]) -> Vec<u8>;
 }
+
+/// SymmetricEncryptorDecryptor encrypts message to cipher and decrypts cipher to message.
+///
+pub trait SymmetricEncryptorDecryptor {
+    /// Encrypts the message returning the tuple (cipher, nonce and padding) if success or error otherwise.
+    ///
+    fn encrypt(&self, message: &[u8]) -> Result<(Vec<u8>, [u8; 32], usize), ErrorSecure>;
+    /// Decrypts returns the plain buffer. Arguments are message and nonce.
+    ///
+    fn decrypt(&self, cipher: &[u8], nonce: &[u8; 32], padding: usize) -> Vec<u8>;
+}
```

Comment:
Suggestion: I'd prefer to separate traits for encryptor and decryptor.

These are skills - I can encrypt or I can decrypt. The struct implementation can have both skills, but not necessarily.
Encrypted message, can also be undecryptable (e.g. hash)

Reply:
Hashing is not encrypting. Hash is a one-way function checksum.

Reply:
Correct. Who knows that message is encrypted or hashed? Only developer who knows that message will be decrypted.

From helicopter view encryption and hashing is the same. It's human-unreadable message.
They differ in usage:
* Hash, you only compare (e.g. password hash compared to stored password hash)
* Encryption, you may decrypt if you know the key

What's more, if it wasn't RSA or AES you'd be able to use encryption as hash if you wanted. Let's say you encrypt message by replacing letter character with a number. You can decrypt, but if you wanted you could also check the "hash" if they are the same.
* In modern encryption tools, public key generates different content because of the salt.

Reply:
Probably we should make more of the traits smaller and then embed smaller traits into the larger ones. Theoretically, it is sound.
I am creating the trait only for the reason to allow other encryption-decryption protocols. They are bound together in a way that you cannot encrypt with one and decrypt with another. In our case, the abstraction is the smallest minimal functionality. Less than that and it doesn't work as expected.

Hashing is not a cryptographic function (a well-designed hashing is) but modulo can create a good enough hash checksum and it isn't good cryptography (it is called then https://en.wikipedia.org/wiki/Cryptographic_hash_function). But let's leave hashing alone for a second.

The rule about a trait being minimal (having minimal possible functionality) doesn't mean that it shall always have one function. For me, it is something saying this trait describes a minimal working abstraction. Having a thing that can only decrypt makes no logical sense.  Decrypt from what (if this can decrypt everything that can be encrypted then I would agree).

We will need that trait probably later. But I would rather stick to one implementation that is strong like AES and restrict usage of other protocols to a bare minimum.

Reply:
On the other hand, you can create one guy with an enigma machine and teach him to encrypt, and the other guy with an enigma machine and teach him only to decrypt. Both will agree on the protocol (the enigma machine). So I think your point is valid. But I will not use two guys or girls for that. I want one that is fully capable of both and I don't want the less than that. It is a weaker abstraction but I want this coupling, so I always have both actions as I will always want both actions. It will probably make more sense to move the trait where I am consuming (will in the future) the abstraction. Then You will see that I will never separate this functionalities.